ENH: Preserve key order when passing list of dicts to DataFrame on py 3.6+

pilkibun opened the pull request:
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Related to #25915,  #26587, #24859, #26113, #10056 (orderedicts), #11181/#11416 (list of namedtuple).  #25911.

**Update**: #13304/#13309 Was merged three years ago so let's just make list of dicts act like list of OrderedDict (as pointed out by @jorisvandenbossche).

**Actual**
```python
In [63]: data= [
    ...: {'name': 'Joe', 'state': 'NY', 'age': 18},
    ...: {'name': 'Jane', 'state': 'KY', 'age': 19}
    ...: ]
    ...: pd.DataFrame(data)
Out[63]: 
   age  name state
0   18   Joe    NY
1   19  Jane    KY
```
**Expected**
```
In [64]: pd.DataFrame(data)
Out[64]: 
   name state  age
0   Joe    NY   18
1  Jane    KY   19
``` 
<del>
Four years ago, #10056 asked for the implied order of columns in a list of `OrderedDict` to be preserved by the `DataFrame` constructor. @thatneat  [commented](https://github.com/pandas-dev/pandas/issues/10056#issuecomment-509383829) yesterday that with 3.7 guaranteed dict order, this should extend to dict-like in general. I think users have a reasonable expectation for this to work, and therefore that pandas should support it. @jreback [voted](https://github.com/pandas-dev/pandas/issues/10056#issuecomment-98812435) +0 on adding this (four years ago).
 
`namedtuple`  has the convenient property of homogeneous keys and key-order which a list of dicts
doesn't have, dicts are allowed to omit keys, and the key order also may change from dict to dict.
Given that, I settled on a reasonable compromise that matches user expectations in practice:
1. Only look at the first dict in the list.
2. Only guarantee the column order of the keys which actually appear in it.
3. **Clarification** the order among columns not included in the first dict is undefined, except that they will appear after all the columns that do.
4. **Added** Changes apply to Python3.6+ only

In practice, I think the only case users actually care about is sensible behavior when passing a list of dicts which is homogeneous in terms of key and key-order, which this PR provides.
</del>

Commit:
ENH: Support new case of implied column ordering in Dataframe()

WillAyd commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1340,6 +1340,31 @@ def test_constructor_list_of_namedtuples(self):
+        data = [
+            {"First": 1, "Second": 4, "Third": 7, "Fourth": 10},
+            {"Second": 5, "First": 2, "Fourth": 11, "Third": 8},
+            {"Second": 6, "First": 3, "Fourth": 12, "Third": 9, "YYY": 14, "XXX": 13},
```
What is the order here if YYY precedes Second? Somewhat wary of making assertions about ordering in 2D

pilkibun replied:
undefined.

thatneat replied:
yeah, it's tricky... as mentioned in my general comment, we could probably use `dict.update()` as a guide. As I understand it the guiding principle is insertion order - whatever is inserted first stays where it is. You don't have to deal with reordering things that were already inserted, just tack the new keys on to the end in the order you get them.

pilkibun replied:
btw, I really appreciate how responsive you are to PRs. Just wanted to say thank you.

thatneat commented:
Thanks for following through on this!
If I can ask for a small amount of scope creep, though - it seems like all other things being equal, the ordering of columns added "later on" should still match the input if possible. This would be analogous to the behavior of `dict.update()`:

```
In [1]: d={5:4, 3:2}                                                                                                 

In [2]: list(d.keys())                                                                                               
Out[2]: [5, 3]

In [3]: d.update({3:3, 5:5, 4:4, 1:1})

In [4]: list(d.keys())
Out[4]: [5, 3, 4, 1]
```

That said, this is already a big improvement and I agree with you that you're covering the most common case. If implementing the above turns out to be too tricky for now, maybe you could just remove the part of the test case that explicitly shows that "added later" `"XXX"` and `"YYY"` columns are sorted - I wouldn't want to imply that that's a desired behavior and close down the option of implementing this later.

pilkibun commented:
@thatneat, I'd like to avoid that. The key order can be different from dict to dict, and there's no clear way to resolve that, which is also simple and doesn't requires looking at all the data.

This solves most of the problem, and I think we should settle for that.

> maybe you could just remove the part of the test case that explicitly shows that "added later" "XXX" and "YYY" columns are sorted

I think you've misread the test. I don't test the order I ignore it by slicing both to the same order explicitly. Solely in order to test  the data values are consistent. This matches the behavior as described in OP.

thatneat commented:
> I think you've misread the test.
yep, I totally did. Never mind then!

Commit:
Safer

Commit:
Restrict to Index case

Commit:
Fix tests

pep8speaks commented:
Hello @pilkibun! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/b7320961599c861be2076527d34295d2b3044bdf/pandas/core/frame.py):

> [Line 450:89](https://github.com/pandas-dev/pandas/blob/b7320961599c861be2076527d34295d2b3044bdf/pandas/core/frame.py#L450): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 88 characters)

Commit:
Style

Commit:
Fix test

Commit:
rename

Commit:
Restrict to PY37

Commit:
Style

Commit:
Restrict to PY36

Commit:
Work around fake test failure on PY35

Commit:
Style

WillAyd commented:
So this has come up a few times in other incarnations but I don't really see the point of this PR. If the data is truly ordered then representing as a list of dicts on the way in is a rather inefficient way of storing the data. If it's not ordered picking the first row or just iterating keys in a 2D plane seems rather arbitrary so not really sure the value add. 

There are a lot of edge cases and nuances that make behavior undefined so I generally don't see a value add

pilkibun commented:
> So this has come up a few times in other incarnations but I don't really see the point of this PR.

which incarnations do you mean? named tuples order is preserved. OrderedDict order is preserved (in another context).

>  If the data is truly ordered then representing as a list of dicts on the way in is a rather 
> inefficient way of  storing the data. 

One record per row is extremely common and users have a reasonable expectation that it should work this way. Inefficiency is important at large scale, but at small-medium scale productivity dominates.

>  If it's not ordered picking the first row or just iterating keys in a 2D plane seems rather arbitrary.

I never considered 2D, I only explained why not. Picking the first row is all you need if the data is homogeneous, and means that doing this costs only constant overhead.

> There are a lot of edge cases and nuances that make behavior undefined 

It requires Python3.6 and In the most common case (homogeneous) it "just works". In other cases (far less common I assume), it works if the first item contains all the data, otherwise we promise nothing. That's not "a lot of edge cases". undefined behavior is not essentially a bad thing, it just means pandas makes no guarantees.

Commit:
fix test

thatneat commented:
> I don't really see the point of this PR

To supply a data point here:

My teams have run in to this issue fairly frequently and have been consistently surprised by/ had to work around `DataFrame` column ordering behaviors.  A common case is in tests where you want it to be extremely easy to construct a DataFrame fixture with a particular shape, but turns out to be more complex because we have to explicitly pass `columns` or use less-straightforward data types than dict. It's not a huge cost to deal with it, but I can attest that this is a fairly common source of friction and this PR will help.

@WillAyd What are the edge cases and nuances you're thinking of?
Does my suggestion of using the maintaining-insertion-order behavior `dict` and `dict.update()` help? It may not always be ordering the user wants, but at least it is consistent and easy to explain.

You may find [this python-dev conversation](https://mail.python.org/pipermail/python-dev/2017-December/151265.html) useful. Python core developers recently (late 2017) decided that they had all of the edge cases nailed down enough to commit to insertion ordering for `dict`s. Now that they've figured it out, maybe it's a natural course for pandas to follow along?

WillAyd commented:
The topic of Python insertion order into a single dictionary is not entirely relevant when talking about a list of dictionaries. I'm not sure why we would assume the first row is really indicative of anything in all of the below cases (should be non-trivial to think of more)

```python
>>> [{'z': 1}, {'c': 1, 'b': 1', 'a': 1]}]
>>> [{'a': 1, 'b': 1, 'c': 1}, {'c': 1, 'b': 1', 'a': 1]}]
>>> [{'a': 1, 'b': 1, 'c': 1}, {'z': 1, 'y': 1', 'x': 1]}]
>>> [{'x': 1, 'y': 1, 'z': 1}, {'a': 1, 'y': 1', 'z': 1]}]
```

pilkibun commented:
> I'm not sure why we would assume the first row is really indicative of anything in all of the below cases 

You're basically saying that in the cases this PR doesn't addresses it doesn't guarantee anything. Exactly.

jreback reviewed (changes requested):
I am -0 on this change because its hard to know the 'rule' for a user and its pretty arbitrary that its the first row's keys (which is of course data dependent).

 it is *much* simpler to simply pass ``columns=`` if you actually do care of about the ordering  or ``.reindex()`` anytime you actually want a specific ordering.

I am not sure there is a 'good' soln here.

jorisvandenbossche commented:
About the "other incarnations" that @WillAyd mentioned that recently came up, see https://github.com/pandas-dev/pandas/issues/26587, https://github.com/pandas-dev/pandas/issues/24859, https://github.com/pandas-dev/pandas/pull/26113 and some others (I though we had a PR doing exactly the same change as this recently, but can't directly find it)

Personally, I would be fine with a change in the line of this PR. But for the me, the main reason is that I think that we should make the treatment of "lists of records" more consistent in general. Whether it are lists of series, lists of dicts or lists of namedtuples, those cases should be broadly equivalent.

- Lists of Series currently preserves column order based on the index. And it seems they try to preserve the order of the first as much as possible, appending new observed ones. 
- Lists of namedtuples only takes the order (and length!) of the first, which leads to some broken cases (eg tests a list of namedtuples with different fields -> they get silently written to wrong columns) 
- Lists of dicts -> the keys are sorted.

So I don't think we should try to think of a new "set or rules" for dicts (as spelled out in the top post). 
But I I would personally welcome a change trying to make this handling of "lists of record-like objects" consistent.

jorisvandenbossche commented:
Opened a separate issue for the unsafe namedtuple behaviour: https://github.com/pandas-dev/pandas/issues/27329

thatneat commented:
> it is much simpler to simply pass columns= if you actually do care of about the ordering or .reindex() anytime you actually want a specific ordering.

As a user, I beg to differ. I regularly see cases (see my previous comments) that would be more straightforward and intuitive if you didn't have to pass `columns` or `reindex`, as judged by seeing my team members try without `columns` or `reindex` first, and then be disappointed when they realize they have to.

Completely agree with @jorisvandenbossche above re: focusing on consistency and not reinventing the wheel. It sounds like Lists of Series already follows the maintain-insertion-order (MIO) rule that I advocated above. Landing on something consistent would be really helpful to me and my colleagues.

The data dependence of ordering is an odd side-effect, but it seems worth it for consistency.

thatneat commented:
I'd love to know if PRs moving things more towards MIO-based unified handling of "lists of record-like objects" would be considered in general. If that direction were agreed upon, this PR seems like one of the more important steps but there would be more that I would like to contribute.

pilkibun commented:
Points by @jreback 

> I am -0 on this change because its hard to know the 'rule' for a user

The rule is, iff you pass a list of like-keyed dicts, pandas does what you expect. That's it. I said little to nothing about other cases (I should really have said nothing), and they can remain undefined behavior just as they are now. The user should then pass `columns`.

> it is much simpler to simply pass columns=

And it's also unnecessary in this case. What is the harm in having pandas behave as users expect in a very common case, and keeping the other cases undefined? the benefit OTOH is obvious.

Points by @jorisvandenbossche 

> I think that we should make the treatment of "lists of records" more consistent in general.

Fair. Just to be clear, is the definition of record "an ordered set of key-value pairs"?

> Lists of dicts -> the keys are sorted.

did you mean sorted (as in lexically) or ordered (what we've been discussing)?

> So I don't think we should try to think of a new "set of rules" for dicts

Let's be precise about the ordering guarantee pandas provides. It's simply
```
If you pass a like-ordered like-keyed list of records, pandas will preserve the key ordering
``
That's it. We don't need to argue about other, undefined behavior.  

That behavior is already true for namedtuples, and this PR makes it true also for dicts. 

But if you want to make further guarantees, "there is no good solution" as @jreback said, meaning you have to make arbitrary choices.

If you want to add the reasonable guarantee that you'll never drop keys, then this PR is what you want, and namedtuples need to be fixed (which I'm happy to tackle after this).

pilkibun commented:
See also #25915.

From #26587

```
WillAyd: There may not be a unique ordering if multiple dicts are passed in.
soerenwolfers: If so, what prevents pandas from using the unique order if it exists?
```
and no more was said. 
To be clear, I'm not criticising @WIllAyd, he's just really active on reviewing PRs (:+1: ), so he pops up a lot. 

And I actually agree with him, in the *general case*. But this is not about the general case. This PR is only about the common case of a list of like-keyed, like-ordered records. In all other cases, pandas requires the user to pass `columns` anyway, so pandas is free to do whatever it likes.


With the issues linked in OP, and those by Joris and this last one, that makes for 6 different users who have asked for this feature. Really, isn't that enough?

jorisvandenbossche commented:
> Just to be clear, is the definition of record "an ordered set of key-value pairs"?

I don't have an exact definition, but something like that yes. I think Series, dicts and namedtuples are all similar enough (and record-like) to expect them to be handled similarly.

> > Lists of dicts -> the keys are sorted
>
> did you mean sorted (as in lexically) or ordered (what we've been discussing)?

I was describing the current behaviour (which is sorted lexically, not random as you said in the top post)

> Let's be precise about the ordering guarantee pandas provides. I'm working under the assumption that It's simply
>
> > If you pass a like-ordered like-keyed list of records, pandas will preserve the key ordering
>
> That's it. We don't need to argue about other, undefined behavior. 
> That rule already holds for namedtuples, and this PR makes it true also for dicts.

That's the basic "rule" for Series, but not the full rule. For Series, other behaviour is not undefined (we don't want undefined behaviour, at least it should be deterministically sorted as is now). 
I would need to look more closely at the actual code, but I *think* the logic for Series is basically doing a "union" of the indices (without sorting, so `idx1.union(idx2, sort=False)` for two indexes, but then of course using another routine that can handle multiple).

> "There is no good solution" as @jreback said, to making any further guarantees on ordering. Because they'll have to be arbitrary in some sense.

There is a perfectly fine solution, it is what Series already does, and the same logic is followed by the Index.union method.

jorisvandenbossche commented:
> There is a perfectly fine solution, it is what Series already does, and the same logic is followed by the Index.union method.

BTW, we already use this exact behaviour when passing OrderedDicts:

```
In [3]: records = [OrderedDict([('c', 1), ('a', 2)]), OrderedDict([('b', 3), ('a', 4)])]                                                                      

In [4]: pd.DataFrame(records)                                                                                                                                 
Out[4]: 
     c  a    b
0  1.0  2  NaN
1  NaN  4  3.0
```

So basically what I want to say: we already have the defined behaviour *and* code for Series and OrderdDict, so I don't see any reason to not do this for normal dicts as well since they are now ordered as well. 
(and I would then separately also fix the namedtuple case, as that feels buggy, but let's discuss that separately)

Given the above, I think this PR can be simplified a lot. I think it is basically making the check for OrderedDict to also allow normal dicts:

https://github.com/pandas-dev/pandas/blob/2d0b20b719452dd2539fbd4ad2c5e48e8c4c95d1/pandas/core/internals/construction.py#L539-L542

pilkibun commented:
> I was describing the current behaviour (which is sorted lexically, not random as you said in the top post)

a choice which predates python3.6/7, it made sense when the key order was randomized. And as you've noted, OrderedDicts order preservation has been supported for a long time. But, you're actually in favor of this (generally speaking) anyway, so ok.

> For Series, other behaviour is not undefined 

Are you absolutely sure about this? I'm not saying it's not *determinstic* (this PR is too), by *undefined behaviour* I'm saying that pandas makes no ordering guarantees about the result. It just so happens that the behavior derives from `index.union`, which has specific ordering guarantees. That does not makes it a guarantee by the pandas API, it's only an implementation detail unless the docstring or documentation promise to do this. And I hope it doesn't, see below.

> but I think the logic for Series is basically doing a "union" of the indices

I don't understand. This Pr doesn't drop any keys under any circumstance.

>  (without sorting, so > idx1.union(idx2, sort=False)

afacit, the series case does what @thatneat has been asking for and I was -1 because of the performance implications. It includes the keys in the order of their appearance in the records,
left-to-right and top-down:
```python
In [283]: s1=pd.Series(range(2), pd.Index(["d","c"]))
     ...: s2=pd.Series(range(4),pd.Index(["b", "a", "c","d"]))
     ...: s3=pd.Series(range(5),pd.Index(["b", "x", "c", "a","d"]))
     ...: pd.DataFrame([s1, s2, s3]).columns
Out[283]: Index(['d', 'c', 'b', 'a', 'x'], dtype='object')
```

> There is a perfectly fine solution, it is what Series already does, and the same logic is followed by the 
> Index.union method.

And this fine solution is to do as many joins as there are rows?

```python
s=pd.Series(range(5),pd.Index(["a", "b", "c", "d","e"]))
coll=[s]*100000
%timeit df=pd.DataFrame(coll) 
```

What's the runtime of that on your computer? on mine it's unaccptable.

Surprisingly, this isn't as bad for dicts (what's wrong with `index.union`???)

```python
s=pd.Index(["b", "x", "c", "a","d"]).to_series()
coll=[dict(a=1,b=2,c=3,d=4)]*100000
%timeit df=pd.DataFrame(coll)
%timeit names={k for d in coll  for k in d}
```

it looks like list of dicts is so slow for the constructor already, that figuring out the keys in order may be the least of our trouble. That's an even stronger guarantee than I was hoping for, and I don't actually need it, but if that'll get merged, I'm fine with it.

pilkibun commented:
> I think this PR can be simplified a lot. I think it is basically making the check for OrderedDict to also 
> allow normal dicts:

No problem, I'll look into it.

Commit:
ENH: treat dict like OrderedDict for PY36+ in dataframe constructor

Commit:
Restore frame.py

pilkibun commented:
> I think this PR can be simplified a lot. I think it is basically making the check for OrderedDict to also
> allow normal dicts:

Well, I'll be darned.

Commit:
restore test_normalize.py

Commit:
Skip some json tests on Py35

Commit:
fix test

Commit:
black

jorisvandenbossche commented:
> > but I think the logic for Series is basically doing a "union" of the indices
>
> I don't understand. This Pr doesn't drop any keys under any circumstance.

Not an intersection, a union (a union doesn't drop keys)

jorisvandenbossche commented:
Regarding the performance: creating a DataFrame from a list of Series is indeed awfully slow. But if you would profile it, you would see that the part of unioning the indexes to obtain the desired column names is only a very tiny part and is not at all the bottleneck. It is apparently the `take` that is done for each Series, something that probably can be optimized if someones wants to take a look at it. 
The performance for a list of dicts seems OK to me

jorisvandenbossche commented on pandas/core/internals/construction.py:
```diff
@@ -538,7 +538,8 @@ def _list_of_series_to_arrays(data, columns, coerce_float=False, dtype=None):
     if columns is None:
         gen = (list(x.keys()) for x in data)
-        sort = not any(isinstance(d, OrderedDict) for d in data)
+        additional = dict if PY36 else OrderedDict
```
maybe do `(dict, OrderedDict) if PY36 else OrderedDict` here and use that as the class to check on the line below?

pilkibun replied:
done

jorisvandenbossche commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1340,6 +1340,31 @@ def test_constructor_list_of_namedtuples(self):
         result = DataFrame(tuples, columns=["y", "z"])
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.skipif(not PY36, reason="Guaranteed dict order is Python>=3.7")
```
Python >= 3.6 ?

pilkibun replied:
fixed

jorisvandenbossche commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1340,6 +1340,31 @@ def test_constructor_list_of_namedtuples(self):
+                "Third": [7, 8, 9],
+                "Fourth": [10, 11, 12],
+                "XXX": [None, None, 13],
+                "YYY": [None, None, 14],
```
I think XXX and YYY need to be switched in order? (to match the input data)

pilkibun replied:
fixed

jorisvandenbossche commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1340,6 +1340,31 @@ def test_constructor_list_of_namedtuples(self):
+        assert set(result.columns) == set(expected.columns)
+        # order of first 4 columns dictated by data[0]
+        tm.assert_frame_equal(result.iloc[:, :4], expected.iloc[:, :4])
+        # the order of the remaining two is arbitrary, but assert the data matches
```
Did you actually observe that this was arbitrary? 
(I already tried to explain it is not)

pilkibun replied:
it was in with my original solution, it's not now but I forgot to update it (and it passed). 

Pushed a fix.

jorisvandenbossche commented on pandas/tests/indexing/test_indexing.py:
```diff
@@ -231,6 +231,7 @@ def test_setitem_dtype_upcast(self):
         df.loc[0, "c"] = "foo"
         expected = DataFrame([{"a": 1, "c": "foo"}, {"a": 3, "b": 2, "c": np.nan}])
+        expected = expected[df.columns]  # compensate dict key ordering
         tm.assert_frame_equal(df, expected)
```
you can pass an argument to assert_frame_equal to ignore the column order (and then I would also only do that for older pythons, and ensure that on recent python actually the correct `expected` is created)

pilkibun replied:
didn't know that. will do.

jorisvandenbossche replied:
Can you do the check_like trick also here?

pilkibun replied:
done

jorisvandenbossche commented on pandas/tests/io/json/test_normalize.py:
```diff
@@ -351,32 +353,35 @@ def test_non_ascii_key(self):
+            b"\xc3\x9cnic\xc3\xb8de".decode("utf8"): [0, 1],
             "sub.A": [1, 3],
             "sub.B": [2, 4],
-            b"\xc3\x9cnic\xc3\xb8de".decode("utf8"): [0, 1],
```
Why is this change needed? (this input is a dict, not list of dicts, and plain dicts should not be impacted here (they also already preserve order))

pilkibun replied:
it definitely failed. My guess is json_normalize calls the list case on dataframe, which changes behavior.

jorisvandenbossche replied:
Ah, that makes sense. 
We might need to be careful here though. Do you know if the dicts from the json reader preserve the order of the file? Probably yes, but if not, we should still sort in this case to not have a random order.

pilkibun replied:
```
In [8]: pd.read_json(json.dumps([dict(a=1,b=2),dict(a=1,b=2, y=1, x=3),dict(b=1,a=2,z=4)]))
Out[8]: 
   a  b    y    x    z
0  1  2  NaN  NaN  NaN
1  1  2  1.0  3.0  NaN
2  2  1  NaN  NaN  4.0
```
I think that's ok. And the set of failing has been just these. If it were random, the many builds per CI, and the several pushes would have flush that out I think.

Commit:
CI

Commit:
clean isinstance check

Commit:
messages

Commit:
Update test after behavior change

Commit:
Ignore column order on py35

pilkibun commented:
> > >        but I think the logic for Series is basically doing a "union" of the indices
> >     I don't understand. This Pr doesn't drop any keys under any circumstance.
> Not an intersection, a union (a union doesn't drop keys)

my point was that this PR does/did the equivalent of a union as well, so in that it isn't different from series behavior. nevermind, we're past that.

jreback commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1340,6 +1340,27 @@ def test_constructor_list_of_namedtuples(self):
+            }
+        )
+        result = DataFrame(data)
+        assert set(result.columns) == set(expected.columns)
```
this line is redudant with the assert_frame_equal

pilkibun replied:
right

Commit:
clean

jorisvandenbossche commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1340,6 +1340,26 @@ def test_constructor_list_of_namedtuples(self):
+            }
+        )
+        result = DataFrame(data)
+        tm.assert_frame_equal(result, expected, check_like=PY36)
```
```suggestion
        tm.assert_frame_equal(result, expected, check_like=not PY36)
```

Its on older versions we need to do the check like?

pilkibun replied:
we want check_like=true on python>=3.6, right? isn't this right?

jorisvandenbossche replied:
I think not. We should ensure that both expected and result are correct (preserving the order) for > py36, and then for older version allow different column order. Am I missing something?

pilkibun replied:
it is now `not PY36` in all cases, which means (IIUC) that `check_like` is True on Py35, and so column order is ignored on py35. 

Really badly named arg.

jreback replied:
> Really badly named arg.

pls cut the commentary

pilkibun replied:
It's confusing and I'm saying so.

jorisvandenbossche replied:
Please @pilkibun, you are free to say it (I often see and mention confusing things in pandas as well), but you can also say it in a more respectful way

pilkibun replied:
I didn't realize it *was* disrespectful or I wouldn't have said it. I'm still not sure why it triggered jreback. No offense intended. I regularly say similar things about code I've written 6 months ago.

jorisvandenbossche reviewed (commented):
Thanks for the updates!

jorisvandenbossche commented:
@WillAyd you're OK now with this change? 
When judging, just forget all long discussions we had before (I think they were partly misguided), but instead look at the clean short diff, and the fact that we already do this for OrderedDict, so we are just extending that to dict with the exact same behaviour as already exists.

WillAyd commented on pandas/tests/io/json/test_normalize.py:
```diff
@@ -366,21 +368,21 @@ def test_missing_field(self, author_missing_data):
         ]
         expected = DataFrame(ex_data)
-        tm.assert_frame_equal(result, expected)
+        tm.assert_frame_equal(result, expected, check_like=PY36)
```
Shouldn't this be `not PY36`?

pilkibun replied:
yes, fixed.

WillAyd reviewed (commented):
Yea looks good. Can you add a whatsnew?

Commit:
fix

Commit:
fix

pilkibun commented:
Could you close #10056 (it was fixed by #13309 long ago)? I'll remove it from the OP "closes #", since this PR doesn't add anything to OrderedDicts which already work.

Commit:
whatsnew

pilkibun commented:
> Can you add a whatsnew?

added. Is this visible enough to put it in the hero section?

Commit:
fix issue ref

jorisvandenbossche commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -181,6 +181,15 @@ The repr now looks like this:
     json_normalize(data, max_level=1)
 
+Column order is preserved when passing a list of dicts to DataFrame
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
We should maybe put it in the API changes section? As some people do rely on column order

pilkibun replied:
done.

jorisvandenbossche commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -181,6 +181,15 @@ The repr now looks like this:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A long requested feature, this has been the behavior for lists of
+`OrderedDict` since v0.19.0. With dict key-order now officially
```
Please keep it a bit "dryer", the "A long requested feature" is not needed (we have many other things which are much more requested)

pilkibun replied:
You have many more important things which users requested, but we've established that this feature has a pretty long request history. Still, point taken.

Commit:
fix header type of unrelated issue

Commit:
whatsnew

Commit:
checks

jorisvandenbossche commented on pandas/tests/indexing/test_indexing.py:
```diff
@@ -231,7 +231,9 @@ def test_setitem_dtype_upcast(self):
 
         df.loc[0, "c"] = "foo"
         expected = DataFrame([{"a": 1, "c": "foo"}, {"a": 3, "b": 2, "c": np.nan}])
```
```suggestion
        expected = DataFrame([{"a": 1, "b": np.nan, "c": "foo"}, {"a": 3, "b": 2, "c": np.nan}])
```

jorisvandenbossche commented on pandas/tests/indexing/test_indexing.py:
```diff
@@ -231,7 +231,9 @@ def test_setitem_dtype_upcast(self):
-        tm.assert_frame_equal(df, expected)
+        # dataframe dict key order is preserved, so these two
+        # sequences of ops lead to different column order
+        tm.assert_frame_equal(df, expected, check_like=True)
```
```suggestion
        tm.assert_frame_equal(df, expected, check_like=not PY36)
```

and then you can remove the comment above

jorisvandenbossche commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,44 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+      ...:     {'name': 'Jane', 'state': 'KY', 'age': 19}
+      ...: ]
+
+*Previous behavior*
```
```suggestion
*Previous behavior* (lexicographically sorted)
```

or something like that

jorisvandenbossche commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,44 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+   0   18   Joe    NY
+   1   19  Jane    KY
+
+*New behavior*
```
```suggestion
*New behavior*: preserve order of the dicts
```

Commit:
Update pandas/tests/indexing/test_indexing.py

Co-Authored-By: Joris Van den Bossche <jorisvandenbossche@gmail.com>

Commit:
Update pandas/tests/indexing/test_indexing.py

Co-Authored-By: Joris Van den Bossche <jorisvandenbossche@gmail.com>

Commit:
Update doc/source/whatsnew/v0.25.0.rst

Co-Authored-By: Joris Van den Bossche <jorisvandenbossche@gmail.com>

Commit:
Update doc/source/whatsnew/v0.25.0.rst

Co-Authored-By: Joris Van den Bossche <jorisvandenbossche@gmail.com>

Commit:
remove comment

Commit:
Checks

Commit:
Add import

Commit:
CI

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,44 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+Column order is preserved when passing a list of dicts to DataFrame
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Starting with Python 3.7 the key-order of `dict` is `guaranteed <https://mail.python.org/pipermail/python-dev/2017-December/151283.html>`_. In practice, this has been true since
```
use double backtikcs around dict, OrderedDict. Reference to DataFrame is ``:class:`DataFrame` ``

pilkibun replied:
ok

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,44 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+
+.. code-block:: ipython
+
+   In [1]: data = [
```
make this an ipython block

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,44 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+
+*New behavior*: preserve order of the dicts
+
+.. code-block:: ipython
```
don't need the prompts; ipython adds these

pilkibun replied:
This is what all other segments in the file do.

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,44 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+Starting with Python 3.7 the key-order of `dict` is `guaranteed <https://mail.python.org/pipermail/python-dev/2017-December/151283.html>`_. In practice, this has been true since
+Python 3.6. The DataFrame constructor now treats a list of dicts in the same way as
+it has a list of `OrderedDict` since v0.19.0. This change applies only when pandas is running
+on python>=3.6 (:issue:`27309`). As a consequence, the column order produced by `DataFrame()`
```
issue ref goes at the end

jreback replied:
python -> Python

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,44 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+      ...:     {'name': 'Jane', 'state': 'KY', 'age': 19}
+      ...: ]
+
+*Previous behavior* (lexicographically sorted)
```
description goes on the next line (same for new section)

jreback commented on pandas/core/internals/construction.py:
```diff
@@ -538,7 +538,8 @@ def _list_of_series_to_arrays(data, columns, coerce_float=False, dtype=None):
 def _list_of_dict_to_arrays(data, columns, coerce_float=False, dtype=None):
     if columns is None:
         gen = (list(x.keys()) for x in data)
```
add a short description of what the guarantees are, here for future readers

jreback reviewed (changes requested):
looks ok, ping . on green.

Commit:
fix tests

Commit:
whatsnew

Commit:
comment

pilkibun commented:
Addressed all comment. Last minute qualm: do you think some users will miss the ability to have sorted keys? right now they can choose between OrderedDict for insertion-order and dict for sorted. That's can be useful too.

Commit:
whatsnew

Commit:
comment

Commit:
checks

jorisvandenbossche commented on pandas/core/internals/construction.py:
```diff
@@ -536,9 +536,17 @@ def _list_of_series_to_arrays(data, columns, coerce_float=False, dtype=None):
 
 
 def _list_of_dict_to_arrays(data, columns, coerce_float=False, dtype=None):
+    """Convert list of OrderedDict to dict
```
```suggestion
    """Convert list of dicts to arrayst
```

the output are arrays? (or did I misinterpret your sentence?)

pilkibun replied:
You're right, I'm just getting impatient.

jorisvandenbossche commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,52 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+The column order now matches the insertion-order of the keys in the ``dict``,
+considering all the records from top to bottom.
+
+.. code-block:: python
```
@jreback meant to say that this should also be a `.. ipython:: python` block, and then in that case the prompts are indeed not needed

pilkibun replied:
And I said that none of the other segments do that, so why this case? 

Maybe I'm misunderstanding. IIUC `code-block` means "please syntax-highlight", while ipython means "please run and include output". 

If that's correct, why would you want this to ever change based on code changes? and if it's done in previous whatsnew (like the one that introduces sorted keys for `Dataframe(dict)` for example, this PR now breaks *that* documentation. And if this is a code-block, why should the previous snippet be executed by ipython and perhaps also break in the future (like kwd changes)?

In my view, whatsnew should be immutable (for code changes) because it's a specification. And
you seemed to think so at one point too #6856.

jorisvandenbossche replied:
> And I said that none of the other segments do that, so why this case?

If you look at the rest of the file, you will see that *all* other code blocks for "new behaviour" use `.. ipython:: python` blocks (and the ones showing old behaviour use `code-block`). 
That's how it is currently done. You can certainly disagree with that habit, but then a discussion about that can happen in a new issue, not here. 

> And you seemed to think so at one point too #6856.

That issue is for *old* whatsnew files.

pilkibun replied:
ok.

pilkibun replied:
> That issue is for *old* whatsnew files.

I looked and the whatsnew for 0.22, 0.23 and 0.24 at least still contain ipython blocks for "new" behavior, so the problem I described is still on point afaict.

jorisvandenbossche replied:
Therefore, the issue you referenced to convert those, is an *open* issue :-)

jorisvandenbossche reviewed (approved):
Last tiny remark, should be good otherwise (sorry for the many back and forth rounds)

Commit:
docstring

Commit:
whatsnew

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,47 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Starting with Python 3.7 the key-order of ``dict`` is `guaranteed <https://mail.python.org/pipermail/python-dev/2017-December/151283.html>`_. In practice, this has been true since
+Python 3.6. The :class:`DataFrame`  constructor now treats a list of dicts in the same way as
```
extra space before constructor

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,47 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+
+Starting with Python 3.7 the key-order of ``dict`` is `guaranteed <https://mail.python.org/pipermail/python-dev/2017-December/151283.html>`_. In practice, this has been true since
+Python 3.6. The :class:`DataFrame`  constructor now treats a list of dicts in the same way as
+it has a list of ``OrderedDict`` since v0.19.0. This change applies only when pandas is running
```
remove the v0.19.0 reference

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -741,6 +741,47 @@ consistent with NumPy and the rest of pandas (:issue:`21801`).
+Starting with Python 3.7 the key-order of ``dict`` is `guaranteed <https://mail.python.org/pipermail/python-dev/2017-December/151283.html>`_. In practice, this has been true since
+Python 3.6. The :class:`DataFrame`  constructor now treats a list of dicts in the same way as
+it has a list of ``OrderedDict`` since v0.19.0. This change applies only when pandas is running
+on Python>=3.6. As a consequence, the column order produced by :class:`DataFrame`
```
This text is awkward and confusing starting with 'This change applies only when pandas is running...' pls simplify.

pilkibun replied:
Joris kindly rewrote that part. If that's not good enough for you either, please suggest a specific revision.

jreback commented on pandas/core/internals/construction.py:
```diff
@@ -536,9 +536,17 @@ def _list_of_series_to_arrays(data, columns, coerce_float=False, dtype=None):
 
 
 def _list_of_dict_to_arrays(data, columns, coerce_float=False, dtype=None):
+    """Convert list of OrderedDict to numpy array
```
to numpy arrays

jreback commented on pandas/core/internals/construction.py:
```diff
@@ -536,9 +536,17 @@ def _list_of_series_to_arrays(data, columns, coerce_float=False, dtype=None):
 def _list_of_dict_to_arrays(data, columns, coerce_float=False, dtype=None):
+    """Convert list of OrderedDict to numpy array
+
+    if `columns` is not passed, column names are inferred from the records
```
can you add a doc-string for the Parameters & Returns, types would be great too

pilkibun replied:
add Parameters & Returns, skipped types

jorisvandenbossche commented:
@jreback this PR has seen enough back and forth on small details, I think we can just merge as is.

jreback commented:
> @jreback this PR has seen enough back and forth on small details, I think we can just merge as is.

I disagree

jorisvandenbossche committed:
doc comments

Commit:
typo

Commit:
whatsnew

Commit:
document parameters

pilkibun commented:
Thanks @jorisvandenbossche for the fixes. I pushed a few more of @jreback's comments.

For the record, the overhead of doing full column discovery is not negligible. I profile it at about 50% slower compared to passing columns explicitly. I think that's acceptable (it has been for `OrderedDict`), but it should be clear moving forward with this for dicts, and in the future for namedtuples too.

```python
In [4]: import pandas as pd
   ...: 
   ...: from typing import NamedTuple
   ...: from collections import namedtuple
   ...: import gc 
   ...: 
   ...: Foo=namedtuple("Foo","a,b,c,d")
   ...: 
   ...: d=dict(a=1,b=2,c=3,d=4)        
   ...: data1=[d]*100000
   ...: %timeit pd.DataFrame(data1, columns=['a','b','c','d']) 
   ...: %timeit pd.DataFrame(data1)
96 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
132 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
```

jorisvandenbossche commented:
> For the record, the overhead of doing full column discovery is not negligible. 

Yes, but note that this has not changed. Also on master, specifying the column names is faster than doing the discovery. The only change is that the union of the dict keys keep the order instead of doing a sort, and that has basically no performance implication.

jorisvandenbossche commented on pandas/core/internals/construction.py:
```diff
@@ -542,6 +542,20 @@ def _list_of_dict_to_arrays(data, columns, coerce_float=False, dtype=None):
+        collection of records (OrderedDict, dict)
+    columns: iterables or None
+    coerce_float : bool
+        If False, don't copy the data if not needed.
```
not sure this is correct?

pilkibun replied:
sorry.

Commit:
remove wrong description

pilkibun commented:
> Yes, but note that this has not changed.

True, but it will for namedtuples which is next. Anyway, I thought providing some numbers is appropriate.

pilkibun commented:
@jorisvandenbossche,thanks for saving this PR after a false-start and for your patience in reviewing.
